core/src/main/c/share/ooo.cpp

```diff
@@ -701,10 +703,579 @@ Java_io_questdb_std_Vect_radixSortLongIndexAscInPlace(JNIEnv *env, jclass cl, jl
+                    reverse_index_format_bytes, src_addresses,
+                    dst_address, reverse_index,
+                    segment_address_ptr, segment_count,
+                    merge_index_format
+            );
+        case 32:
+            return merge_shuffle_fixed_columns_by_rev_index_from_many_addresses<long_256bit>(
+                    reverse_index_format_bytes, src_addresses,
+                    dst_address, reverse_index,
+                    segment_address_ptr, segment_count,
+                    merge_index_format
+            );
+        default:
+            return -1;
+    }
+}
+
+JNIEXPORT jlong JNICALL
+Java_io_questdb_std_Vect_mergeShuffleStringColumnFromManyAddresses(
+        JNIEnv *env,
+        jclass cl,
+        jlong indexFormat,
+        jint dataLengthBytes,
+        jlong srcPrimaryAddresses,
+        jlong srcSecondaryAddresses,
+        jlong dstPrimaryAddress,
+        jlong dstSecondaryAddress,
+        jlong mergeIndex,
+        jlong dstVarOffset,
+        jlong dstVarSize
+) {
+    auto merge_index_address = reinterpret_cast<const index_l *>(mergeIndex);
+    auto src_primary = reinterpret_cast<const char **>(srcPrimaryAddresses);
+    auto src_secondary = reinterpret_cast<const int64_t **>(srcSecondaryAddresses);
+    auto dst_primary = reinterpret_cast<char *>(dstPrimaryAddress);
+    auto dst_secondary = reinterpret_cast<int64_t *>(dstSecondaryAddress);
+    auto dst_var_offset = __JLONG_REINTERPRET_CAST__(int64_t, dstVarOffset);
+    auto row_count = read_row_count(indexFormat);
+    auto index_segment_encoding_bytes = read_segment_bytes(indexFormat);
+    auto format = read_format(indexFormat);
+    auto dst_var_size = __JLONG_REINTERPRET_CAST__(int64_t, dstVarSize);
+
+    if (format != shuffle_index_format && format != dedup_shuffle_index_format) {
+        return -2;
+    }
+
+    int64_t dst_var_end_offset;
+    switch (dataLengthBytes) {
+        case 4:
+            dst_var_end_offset = merge_shuffle_string_column_from_many_addresses<int32_t, 2u>(
+                    index_segment_encoding_bytes, src_primary,
+                    src_secondary, dst_primary, dst_secondary,
+                    merge_index_address, row_count,
+                    dst_var_offset
+            );
+            break;
+        case 8:
+            dst_var_end_offset = merge_shuffle_string_column_from_many_addresses<int64_t, 1u>(
+                    index_segment_encoding_bytes, src_primary,
+                    src_secondary, dst_primary, dst_secondary,
+                    merge_index_address, row_count,
+                    dst_var_offset
+            );
+            break;
+        default:
+            return -1;
+    }
+
+    if (dst_var_end_offset < 0) {
+        // Error occurred, this is error code
+        return dst_var_end_offset;
+    }
+
+    if (dst_var_end_offset - dst_var_offset > dst_var_size) {
+        return -3;
```

Comment:
magic numbers

Reply:
Why should the error code be named if everything we log is the code number?

Reply:
the meaning of -3 is not clear. how does it different from -2 ?

Reply:
It will be logged as a failure with code -3, and the apply will switch to 1 commit at a time mode. -3 is good enough to find this place in the code.